src/item.cpp

```diff
@@ -4132,7 +4132,7 @@ bool item::gunmod_compatible( const item& mod, bool alert ) const
     } else if( get_free_mod_locations( mod.type->gunmod->location ) <= 0 ) {
         msg = string_format( _( "Your %1$s doesn't have enough room for another %2$s mod." ), tname().c_str(), _( mod.type->gunmod->location.c_str() ) );
 
-    } else if( ammo_remaining() > 0 || magazine_current() ) {
+    } else if( ( ammo_remaining() > 0 || magazine_current() ) && ammo_current() != "generic_no_ammo" ) {
```

Comment:
If this fixes it then as far as the games concerned your charged rifle **is** loaded. Can you attach a screenshot of your game (in particular the sidebar that shows health and time) just before a failed attempt to modify the gun.